d3a_interface/device_validator.py

```diff
@@ -24,10 +25,153 @@
 GeneralSettings = ConstSettings.GeneralSettings
 LoadSettings = ConstSettings.LoadSettings
 PvSettings = ConstSettings.PVSettings
+HomeMeterSettings = ConstSettings.HomeMeterSettings
 StorageSettings = ConstSettings.StorageSettings
 CepSettings = ConstSettings.CommercialProducerSettings
 
 
+class DeviceValidator(metaclass=ABCMeta):
+    """Interface for devices' validator classes."""
+
+    @classmethod
+    def validate(cls, **kwargs):
+        """Validate both rate and energy values of the device."""
+        cls.validate_rate(**kwargs)
+        cls.validate_energy(**kwargs)
+
+    @classmethod
+    @abstractmethod
+    def validate_rate(cls, **kwargs):
+        """Validate the rate values of the device."""
+
+    @classmethod
+    @abstractmethod
+    def validate_energy(cls, **kwargs):
+        """Validate the energy values of the device."""
+
+
+class HomeMeterValidator(DeviceValidator):
```

Comment:
For now it is ok, but if we plan to employ it in other strategies then this class will repeat itself. This validator actually is more appropriate for the TemplateStrategyBidUpdater / TemplateStrategyOfferUpdater. The reason why (I guess) you had to implement it in the strategy class, was because these limits are different for each strategy class, and they needed to differentiate. Nothing to do now, just another enhancement topic would be to refactor the other strategies too to use these validation methods. Please leave it for another PR though, since this will delay the review process.

Reply:
Yes, right now we're just using individual functions to validate (e.g. `validate_storage_device(**kwargs)`, `validate_pv_device(**kwargs)` etc.). These are used both inside d3a strategies, and inside GraphQL output objects in d3a-web.
I decided to use a class to begin a gradual refactoring of this validation process. The next step would be to create classes for the other devices too (removing the individual functions). Once that is done, we can try to improve the internal implementation to avoid too much repetition (not sure how yet).